### PR TITLE
Dev 1.4.0 459 remove classifier none

### LINKED ISSUE #459: Replace Classifier None If statements with NotImplemented Warning pytests reports
**Is your feature request related to a problem? Please describe.**
* Currently whenever a test requires a classifier which has not been implemented for a given framework, the tests for that framework are ignored silently.
* Also some tests are too long to run on Travis and make the whole ART test pass the Travis test time limit

**Describe the solution you'd like**
* What we want is to get a detailed report of when such tests are ignored and for what framework specifically.
* Create a new marker to make a test not run on travis

### PR DESCRIPTION
**Pull request by killianlevacher**
# Description

This PR simplifies the test code to deal with fixtures not available for certain combination of framework/requirements.
It also provides a summary after each test run of what tests could not be run due to fixtures missing for such combination of requirements. 

As such this provides ART for the first time with a map  of what is really being test for each framework and what remains to be test if we wish to do so.

It also adds a new marker which enables tests to be skipped in Travis if necessary

Fixes #459

## Type of change

Please check all relevant options.

- [X] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: run_tests.sh [hunk 1]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
 
 # --------------------------------------------------------------------------------------------------------------- TESTS
 
-pytest -q tests/attacks/evasion/ --mlFramework="tensorflow" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/evasion tests"; fi
+mlFrameworkList=("pytorch")
+for mlFramework in "${mlFrameworkList[@]}"; do
+  pytest -q -s tests/attacks/evasion/test_shadow_attack.py --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:11]:
```suggestion
  pytest -q -vv -s tests/attacks/evasion/test_shadow_attack.py --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 2]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/evasion/test_shadow_attack.py"; fi
+
+  pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:14]:
```suggestion
  pytest -q -vv tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 3]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_pytorch.py tests"; fi
 
-pytest -q -s tests/attacks/evasion/test_shadow_attack.py --mlFramework="pytorch" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/evasion/test_shadow_attack.py"; fi
+  pytest -q tests/classifiersFrameworks/test_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:17]:
```suggestion
  pytest -q -vv tests/classifiersFrameworks/test_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 4]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed classifiersFrameworks/test_pytorch.py tests"; fi
+done
 
 mlFrameworkList=("tensorflow" "scikitlearn")
 for mlFramework in "${mlFrameworkList[@]}"; do
-  pytest -q tests/attacks/inference/ --mlFramework=$mlFramework --durations=0
+  pytest -q tests/attacks/inference/ --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:23]:
```suggestion
  pytest -q -vv tests/attacks/inference/ --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 5]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
   if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/inference tests"; fi
 done
 
 mlFrameworkList=("tensorflow")
 for mlFramework in "${mlFrameworkList[@]}"; do
-  pytest -q tests/defences/preprocessor --mlFramework=$mlFramework --durations=0
+  pytest -q tests/defences/preprocessor --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:29]:
```suggestion
  pytest -q -vv tests/defences/preprocessor --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 6]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
   if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor tests"; fi
 
-  pytest -q tests/utils --mlFramework=$mlFramework --durations=0
+  pytest -q tests/utils --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:32]:
```suggestion
  pytest -q -vv tests/utils --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 7]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
   if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed utils tests"; fi
 
-  pytest -q tests/attacks/evasion/ --mlFramework=$mlFramework --durations=0
+  pytest -q tests/attacks/evasion/ --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:35]:
```suggestion
  pytest -q -vv tests/attacks/evasion/ --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 8]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
   if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed attacks/evasion tests"; fi
-done
 
-pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework="pytorch" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_pytorch.py tests"; fi
+  pytest -q tests/classifiersFrameworks/test_tensorflow.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:38]:
```suggestion
  pytest -q -vv tests/classifiersFrameworks/test_tensorflow.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 9]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed classifiersFrameworks/test_tensorflow.py tests"; fi
 
-pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework="tensorflow" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_tensorflow.py tests"; fi
+  pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:41]:
```suggestion
  pytest -q -vv tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 10]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_tensorflow.py tests"; fi
 
-pytest -q tests/classifiersFrameworks/test_pytorch.py  --mlFramework="pytorch" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed classifiersFrameworks/test_pytorch.py tests"; fi
+  pytest -q tests/attacks/evasion/ --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:44]:
```suggestion
  pytest -q -vv tests/attacks/evasion/ --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 11]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed classifiersFrameworks/test_tensorflow.py tests"; fi
 
 
 #NOTE: All the tests should be ran within this loop. All other tests are legacy tests that must be
 # made framework independent to be incorporated within this loop
 mlFrameworkList=("tensorflow" "keras" "pytorch" "scikitlearn" "mxnet" "kerastf")
 for mlFramework in "${mlFrameworkList[@]}"; do
   echo "Running tests with framework $mlFramework"
-  pytest -q tests/estimators/classification/test_deeplearning_common.py --mlFramework=$mlFramework --durations=0
+  pytest -q tests/estimators/classification/test_deeplearning_common.py --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:56]:
```suggestion
  pytest -q -vv tests/estimators/classification/test_deeplearning_common.py --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 12]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
   if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed estimators/classification/test_deeplearning_common.py $mlFramework"; fi
-  pytest -q tests/estimators/classification/test_deeplearning_specific.py --mlFramework=$mlFramework --durations=0
-  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed estimators/classification/test_deeplearning_specific.py $mlFramework"; fi
+  pytest -q tests/estimators/classification/test_deeplearning_specific.py --mlFramework=$mlFramework --skip_travis=True --durations=0
```
Comment (beat-buesser) [inline run_tests.sh:58]:
```suggestion
  pytest -q -vv tests/estimators/classification/test_deeplearning_specific.py --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh (hunk 7 above)
Comment (beat-buesser) [inline run_tests.sh:35]: It looks like all 41 tests are skipped. Is this intended? If no, could you please also check if other test get also skipped unexpectedly.

File: run_tests.sh (hunk 10 above)
Comment (beat-buesser) [inline run_tests.sh:44]: This line seems to be a duplicate of line 35. Also 41 test are all skipped.

File: tests/utils.py [hunk 13]
```diff
@@ -1339,6 +1341,30 @@ def get_tabular_classifier_kr(load_init=True):
     return krc
 
 
+class ARTTestException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+        # self.errors = errors
```
Comment (beat-buesser) [inline tests/utils.py:1347]:
```suggestion
```

File: tests/defences/preprocessor/test_spatial_smoothing_pytorch.py [hunk 14]
```diff
@@ -24,6 +24,7 @@
 import pytest
```
Comment (beat-buesser) [inline tests/defences/preprocessor/test_spatial_smoothing_pytorch.py:24]: Should similar changes as in this file apply to `test_spatial_smoothing_tensorflow.py`?

File: tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py [hunk 15]
```diff
@@ -120,7 +120,3 @@ def test_relation_clip_values_error(self, is_tf_version_2):
             exc_msg = "Invalid 'clip_values': min >= max."
             with pytest.raises(ValueError, match=exc_msg):
                 SpatialSmoothingTensorFlowV2(clip_values=(1, 0))
-
```
Comment (beat-buesser) [inline tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py:123]: Should similar changes as in file `test_spatial_smoothing_pytorch.py` apply to this file?

File: run_tests.sh (hunk 9 above)
Comment (beat-buesser) [inline run_tests.sh:41]:
```suggestion
  pytest -q tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
```

File: run_tests.sh [hunk 16]
```diff
@@ -6,52 +6,57 @@ export TF_CPP_MIN_LOG_LEVEL="3"
...
 
-pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework="pytorch" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_pytorch.py tests"; fi
+  pytest -q tests/classifiersFrameworks/test_tensorflow.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed classifiersFrameworks/test_tensorflow.py tests"; fi
 
-pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework="tensorflow" --durations=0
-if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_tensorflow.py tests"; fi
+  pytest -q tests/defences/preprocessor/test_spatial_smoothing_pytorch.py  --mlFramework=$mlFramework --skip_travis=True --durations=0
+  if [[ $? -ne 0 ]]; then exit_code=1; echo "Failed defences/preprocessor/test_spatial_smoothing_tensorflow.py tests"; fi
```
Comment (beat-buesser) [inline run_tests.sh:42]: Could it be possible that this test is already covered by line 29?

File: run_tests.sh (hunk 5 above)
Comment (beat-buesser) [inline run_tests.sh:29]: The tests for `test_jpeg_compression`, `test_resample`, and `test_spatial_smoothing` are all skipped, but not the ones for `test_mp3_compression` and `test_inverse_gan`. Is this correct?

File: conftest.py [hunk 17]
```diff
@@ -160,79 +179,83 @@ def setup_tear_down_framework(framework):
...
 
-    if framework == "pytorch":
-        import torch
+        if framework == "tensorflow":
+            if not is_tf_version_2:
+                x_tensor = tf.convert_to_tensor(x_train_mnist.reshape(10, 100, 28, 28, 1))
+                y_tensor = tf.convert_to_tensor(y_train_mnist.reshape(10, 100, 10))
+                # tmp = x_train_mnist.shape[0] / default_batch_size
+                # x_tensor = tf.convert_to_tensor(x_train_mnist.reshape(tmp, default_batch_size, 28, 28, 1))
+                # y_tensor = tf.convert_to_tensor(y_train_mnist.reshape(tmp, default_batch_size, 10))
```
Comment (beat-buesser) [inline conftest.py:201]:
```suggestion
```

**Review by beat-buesser — CHANGES_REQUESTED** [review]
Hi @killianlevacher Thank you very much for your pull request! The changes look good, but I have a few concerns about that all tests are still run after these changes. Could you please take a look at the comments above?

File: conftest.py [hunk 18]
```diff
@@ -20,10 +20,16 @@
 import os
 import shutil
 import tempfile
+import warnings
```
Comment (beat-buesser) [inline conftest.py:23]: Unused import
```suggestion
```

File: conftest.py [hunk 19]
```diff
@@ -75,58 +91,61 @@ def _image_dl_estimator_defended(one_classifier=False, **kwargs):
...
+        return classifier, sess
 
     return _image_dl_estimator_defended
 
 
 @pytest.fixture
-def image_dl_estimator_for_attack(image_dl_estimator, image_dl_estimator_defended):
+def image_dl_estimator_for_attack(framework, image_dl_estimator, image_dl_estimator_defended):
     def _image_dl_estimator_for_attack(attack, defended=False, **kwargs):
+        classifier = None
```
Comment (beat-buesser) [inline conftest.py:116]: This variable seem not to be used.
```suggestion
```

File: conftest.py [hunk 20]
```diff
@@ -20,10 +20,16 @@
 import os
 import shutil
 import tempfile
+import warnings
 
+from keras.preprocessing.image import ImageDataGenerator
+from mxnet import gluon
```
Comment (beat-buesser) [inline conftest.py:26]: Please don't use framework imports at the top to allow running single tests without having to install all frameworks, follow the example on `dev_1.4.0`.

File: conftest.py [hunk 21]
```diff
@@ -20,10 +20,16 @@
...
 import numpy as np
 import pytest
 import requests
+import tensorflow as tf
+from torch.utils.data import DataLoader
+import torch
```
Comment (beat-buesser) [inline conftest.py:32]: Please don't use framework imports at the top to allow running single tests without having to install all frameworks, follow the example on `dev_1.4.0`.

File: conftest.py (hunk 20 above)
Comment (killianlevacher) [reply conftest.py:25]: ok yes that's a good point. so should I move all these imports in each method? Actually, that's one thing I had in my mind for a while (not for this PR of course but in the longer term). To deal with this issue, should we not perhaps split the test/util.py file into framework specific test/util_pytorch etc? This way we wouldn't need to place these imports in multiple methods and would enable users to have to install only 1 framework?

File: run_tests.sh (hunk 16 above)
Comment (killianlevacher) [reply run_tests.sh:42]: yes indeed ! :-) And that's very good news it means the number of lines in run_test.sh is finally started to get reduced, Thanks for spotting it

File: tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py (hunk 15 above)
Comment (killianlevacher) [reply tests/defences/preprocessor/test_spatial_smoothing_tensorflow.py:123]: yes indeed. Thanks for pointing it out, I must have skipped that file

File: conftest.py (hunk 20 above)
Comment (beat-buesser) [reply conftest.py]: I'm not sure of it is required at this stage, for 1.4.0 we have already fixed `conftest.py` and this PR would reintroduce these imports at the top, therefore I would suggest to follow the example on `dev_1.4.0` for now.

**Review by beat-buesser — APPROVED** [review]
Hi @killianlevacher Thank you very much for implementing essential improvements of ART's testing!

**Pull request merged**